README.md
# argonone
Argon One Service and Control Scripts por Raspberry Pi 4

Argon Script delivered with the device:
https://download.argon40.com/argon1.sh

* Creator of PKGBbuild and service files => https://github.com/kounch/argonone
* Modification for RPI4 => https://github.com/Elrondo46/argonone


## Changelog
My contribution is the deb file for debian systems.
* We moved from `/opt` to `/usr/share` for DebianReasons
* renaimed all files to argonone (missing d) so packagebuilder recognizes service file.
* A silent default config is included as default. The difference between `3` and `100` as a fanspeed value in `/etc/argonone.conf` is nearly not noticable. GO Exponential!
* Changed vcgencmd path to `/usr/bin/vcgencmd` - Maybe that's not correct for raspian; but it is for [RPI23](https://github.com/drtyhlpr/rpi23-gen-image)
* Removed python venv as it caused unsolvable trouble for me. As a plus we now have apt updating our libs and don't need to build python packages on install. It's still possible that the advantage of venv negates the positive effects.

## Building Package
Starting from a minimal debootstrap aarch64

1. `apt install git build-essential debhelper`
2. `git clone https://github.com/burnbabyburn/argonone.git && cd argonone && dpkg-buildpackage`
4. `dpkg -i ../*.deb`


debian/changelog
argonone (0.0-1) unstable; urgency=medium

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>
  * We moved from `/opt` to `/usr/share` for DebianReasons
  * renaimed all files to argonone (missing d) so packagebuilder recognizes service file.
  * A silent default config is included as default. The difference between `3` and `100` as a fanspeed value in `/etc/argonone.conf` is nearly not noticable. GO Exponential!
  * Changed vcgencmd path to `/usr/bin/vcgencmd` - Maybe that's not correct for raspian; but it is for [RPI23](https://github.com/drtyhlpr/rpi23-gen-image)

 -- root <root@unknown>  Thu, 30 Apr 2020 21:51:44 +0200


argonone/argononed.py
#!/opt/argonone/bin/python3
# -*- coding: utf-8 -*-
# -*- mode: Python; tab-width: 4; indent-tabs-mode: nil; -*-
# PEP 8, PEP 263.
"""
Argon One Fan and Button Service Daemon
"""

import smbus
import RPi.GPIO as GPIO
import os
import time
from threading import Thread

rev = GPIO.RPI_REVISION
if rev == 2 or rev == 3:
    bus = smbus.SMBus(1)
else:
    bus = smbus.SMBus(0)

GPIO.setwarnings(False)
GPIO.setmode(GPIO.BCM)
shutdown_pin = 4
GPIO.setup(shutdown_pin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)


def shutdown_check():
    while True:
        pulsetime = 1
        GPIO.wait_for_edge(shutdown_pin, GPIO.RISING)
        time.sleep(0.01)
        while GPIO.input(shutdown_pin) == GPIO.HIGH:
            time.sleep(0.01)
            pulsetime += 1
        if pulsetime >= 2 and pulsetime <= 3:
            os.system("reboot")
        elif pulsetime >= 4 and pulsetime <= 5:
            os.system("shutdown now -h")


def get_fanspeed(tempval, configlist):
    for curconfig in configlist:
        curpair = curconfig.split("=")
        tempcfg = float(curpair[0])
        fancfg = int(float(curpair[1]))
        if tempval >= tempcfg:
            return fancfg
    return 0


def load_config(fname):
    newconfig = []
    try:
        with open(fname, "r") as fp:
            for curline in fp:
                if not curline:
                    continue
                tmpline = curline.strip()
                if not tmpline:
                    continue
                if tmpline[0] == "#":
                    continue
                tmppair = tmpline.split("=")
                if len(tmppair) != 2:
                    continue
                tempval = 0
                fanval = 0
                try:
                    tempval = float(tmppair[0])
                    if tempval < 0 or tempval > 100:
                        continue
                except:
                    continue
                try:
                    fanval = int(float(tmppair[1]))
                    if fanval < 0 or fanval > 100:
                        continue
                except:
                    continue
                newconfig.append("{:5.1f}={}".format(tempval, fanval))
        if len(newconfig) > 0:
            newconfig.sort(reverse=True)
    except:
        return []
    return newconfig


def temp_check():
    fanconfig = ["65=100", "60=55", "55=10"]
    tmpconfig = load_config("/etc/argononed.conf")
    if len(tmpconfig) > 0:
        fanconfig = tmpconfig
    address = 0x1a
    prevblock = 0
    while True:
        temp = os.popen("/usr/bin/vcgencmd measure_temp").readline()
        temp = temp.replace("temp=", "")
        val = float(temp.replace("'C", ""))
        block = get_fanspeed(val, fanconfig)
        if block < prevblock:
            time.sleep(30)
        prevblock = block
        try:
            bus.write_byte_data(address,0,block)
        except IOError:
            temp = ""
        time.sleep(30)


try:
    t1 = Thread(target=shutdown_check)
    t2 = Thread(target=temp_check)
    t1.start()
    t2.start()
except:
    t1.stop()
    t2.stop()
    GPIO.cleanup()


argonone/argonone.py
#!/opt/argonone/bin/python3
# -*- coding: utf-8 -*-
# -*- mode: Python; tab-width: 4; indent-tabs-mode: nil; -*-
# PEP 8, PEP 263.
"""
Argon One Fan and Button Service Daemon
"""

import smbus
import RPi.GPIO as GPIO
import os
import time
from threading import Thread

rev = GPIO.RPI_REVISION
if rev == 2 or rev == 3:
    bus = smbus.SMBus(1)
else:
    bus = smbus.SMBus(0)

GPIO.setwarnings(False)
GPIO.setmode(GPIO.BCM)
shutdown_pin = 4
GPIO.setup(shutdown_pin, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)


def shutdown_check():
    while True:
        pulsetime = 1
        GPIO.wait_for_edge(shutdown_pin, GPIO.RISING)
        time.sleep(0.01)
        while GPIO.input(shutdown_pin) == GPIO.HIGH:
            time.sleep(0.01)
            pulsetime += 1
        if pulsetime >= 2 and pulsetime <= 3:
            os.system("reboot")
        elif pulsetime >= 4 and pulsetime <= 5:
            os.system("shutdown now -h")


def get_fanspeed(tempval, configlist):
    for curconfig in configlist:
        curpair = curconfig.split("=")
        tempcfg = float(curpair[0])
        fancfg = int(float(curpair[1]))
        if tempval >= tempcfg:
            return fancfg
    return 0


def load_config(fname):
    newconfig = []
    try:
        with open(fname, "r") as fp:
            for curline in fp:
                if not curline:
                    continue
                tmpline = curline.strip()
                if not tmpline:
                    continue
                if tmpline[0] == "#":
                    continue
                tmppair = tmpline.split("=")
                if len(tmppair) != 2:
                    continue
                tempval = 0
                fanval = 0
                try:
                    tempval = float(tmppair[0])
                    if tempval < 0 or tempval > 100:
                        continue
                except:
                    continue
                try:
                    fanval = int(float(tmppair[1]))
                    if fanval < 0 or fanval > 100:
                        continue
                except:
                    continue
                newconfig.append("{:5.1f}={}".format(tempval, fanval))
        if len(newconfig) > 0:
            newconfig.sort(reverse=True)
    except:
        return []
    return newconfig


def temp_check():
    fanconfig = ["65=100", "60=55", "55=10"]
    tmpconfig = load_config("/etc/argonone.conf")
    if len(tmpconfig) > 0:
        fanconfig = tmpconfig
    address = 0x1a
    prevblock = 0
    while True:
        temp = os.popen("/usr/bin/vcgencmd measure_temp").readline()
        temp = temp.replace("temp=", "")
        val = float(temp.replace("'C", ""))
        block = get_fanspeed(val, fanconfig)
        if block < prevblock:
            time.sleep(30)
        prevblock = block
        try:
            bus.write_byte_data(address,0,block)
        except IOError:
            temp = ""
        time.sleep(30)


try:
    t1 = Thread(target=shutdown_check)
    t2 = Thread(target=temp_check)
    t1.start()
    t2.start()
except:
    t1.stop()
    t2.stop()
    GPIO.cleanup()
